Submodule for UI tests

dkcumming opened the pull request:
This PR adds the [rust compiler](https://github.com/rust-lang/rust) as a submodule, so that the UI tests may be updated. 
- A `doit` script will update the submodule (be aware this is a large download, I couldn't get sparse checkout to work without error) and will update all the tests in our `ui/` directory - if the test is removed, moved, or renamed it will be deleted from `ui/` directory. 
- The Makefile has been simplified, so that tests targeted by previous compilers are removed.
- After updating some tests were unable to compile to mir due to changes in the compiler (also the point above), I removed those tests. This means that the number of tests in `ui/` has changed from ~3500 to ~2700.

dkcumming commented on Makefile:
```diff
@@ -1,12 +1,16 @@
 clean: clean-mir
 
+RUSTC:=rustc
+RUSTC_OPTIONS=-C overflow-checks=off -Zmir-enable-passes=-ConstDebugInfo
```
@yanliu18 Do you feel these flags are sufficent? I notice in handwritten tests there are more

yanliu18 commented:
Regarding the changes in thi PR, it is pretty slow to load in safari (maybe it’s too large). Some further comments, I’ll leave it here.
1. For the not able to compile tests using the 1.72 nightly version, is there a way to ignore them using commands? In case, we needed to add them back in future.
2. Regarding the flags, it is better to be at least consistent with the makefile in handwritten rust tests. You are welcome to try out the other options. One rule is that, we'd better turn off the optimization passes which generates syntax KMIR is unable to parse. 
3. The `ConstDebugInfo` seems not working on some cases. In some examples, the syntax `debug _1 = const A` still exists. I think we do not support `debug` var to be assigned with `const`. Would you experiment on an example to understand why?
4. In the makefile, the commands to compile rust program with older versions, shall we comment out them instead of delete? We might in future need to look into those. What do you think?

dkcumming commented:
Yes it is a lot to load. I think you can improve the performance of the browser by holding `Alt` and clicking the `inverted carrot` symbol, and that should close all windows. And then up the top you can filter out `.mir`, `.rs` files etc.

dkcumming commented:
1. We can bring them all the failing tests back and change the command to compile from
`$(RUSTC) --emit mir $(RUSTC_OPTIONS) -o $@ ui/$*.rs` to
`-$(RUSTC) --emit mir $(RUSTC_OPTIONS) -o $@ ui/$*.rs`
This should stop blocking if there is a compilation error and continue the `make` command with the next file.

2. Okay I will change to be the same as the handwritten and will bring up if I find something different that is interesting.

3. I think I forgot to build with that flag added. The files in the repo right now are without that flag. I will compile with the handwritten flags and push again.

4. We can comment them out. They will be in the version control system but they can get easily forgotten that way.

yanliu18 commented:
> * We can bring them all the failing tests back and change the command to compile from
>   `$(RUSTC) --emit mir $(RUSTC_OPTIONS) -o $@ ui/$*.rs` to
>   `-$(RUSTC) --emit mir $(RUSTC_OPTIONS) -o $@ ui/$*.rs`
>   This should stop blocking if there is a compilation error and continue the `make` command with the next file.
> * Okay I will change to be the same as the handwritten and will bring up if I find something different that is interesting.
> * I think I forgot to build with that flag added. The files in the repo right now are without that flag. I will compile with the handwritten flags and push again.
> * We can comment them out. They will be in the version control system but they can get easily forgotten that way.

Thanks, these should do. As for the 4th issue, let's keep this for a while until some point, we are sure we won't need them any more.

dkcumming commented:
@yanliu18 The compiled mir matches the Makefile now

yanliu18 commented:
> @yanliu18 The compiled mir matches the Makefile now

Thanks, the generated mir file looks well-formatted now.
I guess we should not download rust as submodule (I believe we will be paying for storage, is it correct @ehildenb ? ) 

We might better update the tests locally (by comparing to the lastest test suite in Rust repo) and only upload the changes. What do you think?

dkcumming commented:
Oh does it actually store the module? I thought it was just a pointer to the rust repo essentially? 

It's still overkill to download the whole rust repo to just update the tests, but it shouldn't have to happen much and by many people. 

I did have `sparse-checkout`, but I couldn't get it to work properly.

ehildenb commented:
It also seems to me that adding rustc as a submodule is unnecessary, just provide instructions in teh README about how to obtain the correct version. If you want to use the rustc on CI, you should add a file `deps/rustc`, which contains the version of the `rustc` compiler to use, so that you cna do something like `rustup install $(cat deps/rustc)`, and then it will get you th ecorrect rustc to use. We can use that file on CI as well.

dkcumming commented:
@yanliu18 Seems like the way forward is to compile the tests with a particular `nightly` version, and have that version as the target for `kmir` for the forseeable future. I removed the submodule and added instructions for how to install this particular version of the toolchain. 

My only thoughts are is there any chance that targetting a different version of the toolchain could be preferable? This is just the version from yesterday or so, but maybe there is a preferable version from before. Worth mentioning is that I checked the difference in compilation between a target from last week to this week, and there actually was a breaking difference for some tests (a change of name from `Pointer` to `PointerCoercion`). I wonder how frequently they make changes.

yanliu18 commented:
@dkcumming Indeed, the MIR syntax and its optimization passes are changing from time to time, even in the stable rustc. Th nightly version has lots of experiment features, it would change more frequently. However such changes are not necessary permanent. 

We will need to fix a version until we have the main semantics defined. 
I would prefer a nightly version before a stable release, though I do not have a good idea of how to choose a version. I also think it is not an issue since we will eventually trace back to its unoptimised version and consume the data from memory directly.

I would also expect recompiling the MIR files for test when we discover some other optimization should be turned off.

Lastly, removed the submodule is good. Otherwise the CI will have to download all submodules which is not necessary at all. 
This PR should be good to go.

Thanks.

yanliu18 commented:
@ehildenb Thank you for suggesting the configurations, we will need such a solution in future.

yanliu18 commented:
@dkcumming BTW, in future, when we will need to handle the syntax differences between rustc version, we could implement configuration files/ separate modules for each version.
